src/main/java/org/logstash/beats/BeatsParser.java

```diff
@@ -0,0 +1,230 @@
+                int fieldsCount = (int) in.readUnsignedInt();
+                int count = 0;
+
+                Map dataMap = new HashMap<String, String>();
+
+                while(count < fieldsCount) {
+                    int fieldLength = (int) in.readUnsignedInt();
+                    String field = in.readBytes(fieldLength).toString(Charset.forName("UTF8"));
+
+                    int dataLength = (int) in.readUnsignedInt();
+                    String data = in.readBytes(dataLength).toString(Charset.forName("UTF8"));
+
+                    dataMap.put(field, data);
+
+                    count++;
+                }
+
+                Message message = new Message(sequence, dataMap);
+                batch.addMessage(message);
+
+                if(batch.complete()) {
+                    out.add(batch);
+                    batchComplete();
+                }
+
+                transitionToReadHeader();
+
+                break;
+            }
+            case READ_JSON_HEADER: {
+                logger.debug("Running: READ_JSON_HEADER");
+
+                sequence = (int) in.readUnsignedInt();
+                int jsonPayloadSize = (int) in.readUnsignedInt();
+
+                transition(States.READ_JSON, jsonPayloadSize);
+                break;
+            }
+            case READ_COMPRESSED_FRAME_HEADER: {
+                logger.debug("Running: READ_COMPRESSED_FRAME_HEADER");
+
+                transition(States.READ_COMPRESSED_FRAME, in.readUnsignedInt());
+                break;
+            }
+
+            case READ_COMPRESSED_FRAME: {
+                logger.debug("Running: READ_COMPRESSED_FRAME");
+
+
+                byte[] bytes = new byte[(int) requiredBytes];
+                in.readBytes(bytes);
+
+                InputStream inflater = new InflaterInputStream(new ByteArrayInputStream(bytes));
+                ByteArrayOutputStream decompressed = new ByteArrayOutputStream();
+
+
+                byte[] chunk = new byte[CHUNK_SIZE];
+                int length = 0;
+
+                while ((length = inflater.read(chunk)) > 0) {
+                    decompressed.write(chunk, 0, length);
+                }
+
+                inflater.close();
+                decompressed.close();
+
+                transitionToReadHeader();
+                ByteBuf newInput = Unpooled.wrappedBuffer(decompressed.toByteArray());
+
+                try {
+                    while (newInput.readableBytes() > 0) {
+                        decode(ctx, newInput, out);
+                    }
+                } finally {
+                    newInput.release();
+                }
+
+                break;
+            }
+            case READ_JSON: {
+                logger.debug("Running: READ_JSON");
+
+                byte[] bytes = new byte[(int) requiredBytes];
+                in.readBytes(bytes);
+                Message message = new Message(sequence, (Map) JsonUtils.mapper.readValue(bytes, Object.class));
```

Comment:
Thats a pretty good question, in the current implementation the following will happen:

- An exception will be raised
- The handler will catch it log it.
-  the whole frame will be discarded
- This is what is currently happening in the current ruby implementation.


So this might not be ideal, logstash doesn't have the concept of a DLQ. The protocol does ack on batch so If one of the event cannot be processed we cannot ack the batch.